test/load-generator/acme/directory.go

```diff
@@ -0,0 +1,208 @@
+// server directory URL that is missing a key for a required endpoint in the
+// response JSON. See also RequiredEndpoints.
+type ErrMissingEndpoint struct {
+	endpoint Endpoint
+}
+
+// Error returns the error message for an ErrMissingEndpoint error.
+func (e ErrMissingEndpoint) Error() string {
+	return fmt.Sprintf(
+		"directoryURL JSON was missing required key for %q endpoint",
+		e.endpoint,
+	)
+}
+
+// ErrInvalidEndpointURL is an error returned if NewDirectory is provided an
+// ACME server directory URL that has an invalid URL for a required endpoint.
+// See also RequiredEndpoints.
+type ErrInvalidEndpointURL struct {
+	endpoint Endpoint
+	value    string
+}
+
+// Error returns the error message for an ErrInvalidEndpointURL error.
+func (e ErrInvalidEndpointURL) Error() string {
+	return fmt.Sprintf(
+		"directoryURL JSON had invalid URL value (%q) for %q endpoint",
+		e.value, e.endpoint)
+}
+
+// Directory is a type for holding URLs extracted from the ACME server's
+// Directory resource.
+//
+// See RFC 8555 Section 7.1.1 "Directory".
+//
+// Its public API is read-only and therefore it is safe for concurrent access.
+type Directory struct {
+	// endpointURLs is a map from endpoint name to
+	endpointURLs map[Endpoint]*url.URL
+}
+
+// getRawDirectory validates the provided directoryURL and makes a GET request
+// to fetch the raw bytes of the server's directory resource. If the URL is
+// invalid, if there is an error getting the directory bytes, or if the HTTP
+// response code is not 200 an error is returned.
+func getRawDirectory(directoryURL string) ([]byte, error) {
+	if directoryURL == "" {
+		return nil, ErrEmptyDirectory
+	}
+
+	if _, err := url.Parse(directoryURL); err != nil {
+		return nil, ErrInvalidDirectoryURL
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout:   10 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			TLSHandshakeTimeout: 5 * time.Second,
+			TLSClientConfig: &tls.Config{
+				// Bypassing CDN or testing against Pebble instances can cause
+				// validation failures. For a **test-only** tool its acceptable to skip
+				// cert verification of the ACME server's HTTPs certificate.
+				InsecureSkipVerify: true,
+			},
+			MaxIdleConns:    1,
+			IdleConnTimeout: 15 * time.Second,
+		},
+		Timeout: 10 * time.Second,
+	}
+
+	resp, err := httpClient.Get(directoryURL)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrInvalidDirectoryHTTPCode
+	}
+
+	rawDirectory, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return rawDirectory, nil
+}
+
+// NewDirectory creates a Directory populated from the ACME directory resource
+// returned by a GET request to the provided directoryURL. An error is returned
+// if the directory URL is invalid, or if there is a problem connecting to the
```

Comment:
In general I would recommend against listing all the reasons a function can return an error. It's the kind of comment that's likely to go out of date, and I think it's not the sort of information that someone reading the godoc needs to know. For instance, they're likely to assume that an unparseable URL results in an error.

There's one thing here that's not quite as obvious: checking against RequiredEndpoints. I might write this as "It also checks that the fetched directory contains RequiredEndpoints."

Reply:
Ok

Reply:
Fixed in b330476